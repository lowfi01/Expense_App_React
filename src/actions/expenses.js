import uuid from 'uuid';

// Add Expense
export const addExpense = (
  {
      description = '',
      note = '',
      amount = 0,
      createdAt = 0
  } = {}
) => ({
  type: 'ADD_EXPENSE',
  expense: {
    id: uuid(),
    description,
    note,
    amount,
    createdAt
  }
})


// Remove Expense
export const removeExpense = ({id}) => ({
  type: 'REMOVE_EXPENSE',
  id
})


// Edit Expense
export const editExpense = (id, updates ) => ({
  type: 'EDIT_EXPENSE',
  id,
  updates
});
