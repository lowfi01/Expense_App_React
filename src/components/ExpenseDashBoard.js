import React from 'react';

import ExpenseList from './ExpenseList';
import ExpenseListFilters from './ExpenseListFilters';

export const ExpenseDashBoard = (props) => (
    <div>
      <h1>Expense DashBoard</h1>
      <ExpenseListFilters />
      <ExpenseList />
    </div>

  )

export default ExpenseDashBoard;