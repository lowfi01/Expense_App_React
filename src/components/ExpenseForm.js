import React from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/lib/css/_datepicker.css';

// // use moment
// const now = moment().format('MMMM Do, YYYY, h:mm:ss a');
// console.log(now);


class ExpenseFrom extends React.Component{
  constructor(props) {
    super(props);

    this.state = {
        description: props.expense ? props.expense.description : '',
        note: props.expense ? props.expense.note : '',
        amount: props.expense ? (props.expense.amount / 100).toString() : '',
        createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
        calenderFocused: false,
        error: ''
    }
  }

  onDescriptionChange = (e) => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  }

  onNoteChange = (e) => {
    const note = e.target.value;
    this.setState(() => ({
      note
    }));
  }

  onAmountChange = (e) => {
    const amount = e.target.value;
    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      this.setState(() => ({
        amount
      }))
    }
  }

  // gets created with the moment object
  onDateChange = (createdAt) => {
    if (createdAt) {
      this.setState( () => ({createdAt}));
    }

  }

  onFocuseChange = ({ focused }) => {
    this.setState(() => ({
      calenderFocused: focused
    }))
  };

  onSubmit = (e) => {
    e.preventDefault();

    if (!this.state.description || !this.state.amount) {
      this.setState(() => ({
        error: 'Please provide description and amount'
      }))
    } else {
      this.setState(() => ({
        error: ''
      }))

      // Add validation layer, before sending state object to the parent
      this.props.onSubmit({
        description: this.state.description,
        note: this.state.note,
        amount:  parseFloat(this.state.amount, 10) * 100,
        createdAt:  this.state.createdAt.valueOf(),
      });
    }

  }

  render() {
    return (
      <div>
        { this.state.error && <p>{this.state.error}</p> }
        <form onSubmit={this.onSubmit}>
          <input
            type="text"
            placeholder="Description"
            autoFocus
            value={this.state.description}
            onChange={this.onDescriptionChange}
          />
          <input
            type="text"
            placeholder="Amount"
            value={this.state.amount}
            onChange={this.onAmountChange}
          />
          <SingleDatePicker
            date={this.state.createdAt}
            onDateChange={this.onDateChange}
            focused={this.state.calenderFocused}
            onFocusChange={this.onFocuseChange}
            numberOfMonths={1}
            isOutsideRange={(day) => false}
          />
          <textarea
            placeholder="Add a note for your expense (optional)"
            value={this.state.note}
            onChange={this.onNoteChange}
            />
          <button type="submit" >Submit</button>
        </form>
      </div>
    )
  }
}

export default ExpenseFrom;