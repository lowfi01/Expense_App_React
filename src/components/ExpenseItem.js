import React from 'react';
import { Link } from 'react-router-dom';

export const ExpenseItem = ({id, description, amount, createdAt, dispatch}) => (
    <div>
      <Link to={`/edit/${id}`}>
        <h3>{description}</h3>
      </Link>
      <p>{amount}-{createdAt}</p>

    </div>
  )

export default ExpenseItem;