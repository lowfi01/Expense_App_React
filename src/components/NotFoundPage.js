import React from 'react';
import { Link } from 'react-router-dom';

export const NotFoundPage = () => (
  <div>
    404 - <Link to="/">Go home</Link>
  </div>
)

export default NotFoundPage;