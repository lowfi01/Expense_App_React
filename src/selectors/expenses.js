/* eslint-disable */



import moment from 'moment';

export default (expenses, {text, sortBy, startDate, endDate}) =>
  // remember the combineReducers, uses key value pairs for the object it makes
  // we have these two reducers in our store
  // {
  //  filters: filterReducer,
  //  expenses: expensesReducer
  // }

    expenses.filter((expense) => {
      // const startDateMatch = typeof startDate !== 'number' || expense.createdAt >= startDate ;
      // const endDateMatch = typeof endDate !== 'number' || expense.createdAt <= endDate ;

      const createdAtMoment = moment(expense.createdAt);
      const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
      const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
      const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
   }).sort((a, b) => { // eslint-disable-line

    if(sortBy === 'date') {
      return  a.createdAt < b.createdAt ? 1 : -1;x
     } else if (sortBy === 'amount') {
      return a.amount < b.amount ? 1 : -1;
     }

    })  // sort by date or amount, depending on filter.sortBy definition


// export default getVisibleExpenses;