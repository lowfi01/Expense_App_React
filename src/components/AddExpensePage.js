import React from 'react';
import {connect} from 'react-redux';
import ReactGA from 'react-ga';
import ExpenseForm from './ExpenseForm';

// import { addExpense } from '../actions/expenses';
import { startAddExpense } from '../actions/expenses';

const AddExpensePage = (props) => {
  const onSubmit = (expense) => {
      ReactGA.event({
        category: 'Create Expense',
        action: 'click button',
        label: 'expenses'
      });
      props.dispatch(startAddExpense(expense));
      props.history.push('/');
    }
  return(
      <div>
        <div className="page-header">
          <div className="content-container">
            <h1 className="page-header__title">Add Expense</h1>
          </div>
        </div>

        <div className="content-container">
          <ExpenseForm onSubmit={onSubmit} />
        </div>

      </div>
    )
  }

export default connect()(AddExpensePage);