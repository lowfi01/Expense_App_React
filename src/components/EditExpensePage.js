import React from 'react';
import { connect } from 'react-redux';

import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

const EditExpensePage = (props) => {
  console.log(props);
  return (
    <div>
      <ExpenseForm
        expense={props.expense}
        onSubmit={(expense) => {
              props.dispatch(editExpense(props.match.params.id, expense));
              props.history.push('/');
        }}
      />
      <button onClick={(e) => {
        e.preventDefault();
        props.dispatch(removeExpense({ id : props.expense.id}));
        props.history.push('/');
      }}  >Remove Expense</button>
    </div>
  );
};

const mapStateToProps = (state, props) => ({
    expense: state.expenses.find((expense) => expense.id === props.match.params.id)
  })

export default connect(mapStateToProps)(EditExpensePage);
