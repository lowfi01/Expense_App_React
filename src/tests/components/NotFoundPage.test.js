import React from 'react';
import { shallow } from 'enzyme';
import expenses from '../fixtures/expenses';
import { NotFoundPage } from '../../components/NotFoundPage';

test('should render a not found page', () => {
  const wrapper = shallow(<NotFoundPage />);
  expect(wrapper).toMatchSnapshot();
})
