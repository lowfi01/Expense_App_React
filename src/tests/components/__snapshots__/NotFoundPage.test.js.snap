// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a not found page 1`] = `
<div>
  404 - 
  <Link
    replace={false}
    to="/"
  >
    Go home
  </Link>
</div>
`;
